merkle/smt/writer.go

```diff
@@ -0,0 +1,180 @@
+// Copyright 2019 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package smt
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/google/trillian/merkle/hashers"
+	"github.com/google/trillian/storage/tree"
+)
+
+// NodeBatchAccessor reads and writes batches of Merkle tree node hashes.
+type NodeBatchAccessor interface {
+	// Get returns the hashes of the given nodes, as the corresponding values of
+	// the passed-in map. The initial map values must be nil, and will remain nil
+	// for nodes that root empty subtrees.
+	Get(nodes map[tree.NodeID2][]byte) error
+	// Set applies the given node hash updates.
+	Set(upd []NodeUpdate) error
+}
+
+// Writer handles sharded writes to a sparse Merkle tree.
+type Writer struct {
+	treeID int64
+	hasher hashers.MapHasher
+	shards map[uint]shardWriter
```

Comment:
Add a comment about how the shards are indexed. 
It appears from below that the integer is the height in the tree?

Reply:
Done.